File: tests/algorithms/x/Java/maths/entropy.java
public class Main {
    static class TextCounts {
        java.util.Map<String,Long> single;
        java.util.Map<String,Long> double_;
        TextCounts(java.util.Map<String,Long> single, java.util.Map<String,Long> double_) {
            this.single = single;
            this.double_ = double_;
        }
        TextCounts() {}
        @Override public String toString() {
            return String.format("{'single': %s, 'double': %s}", String.valueOf(single), String.valueOf(double_));
        }
    }

    static String text1;
    static String text3;

    static double log2(double x) {
        double k = (double)(0.0);
        double v_1 = (double)(x);
        while ((double)(v_1) >= (double)(2.0)) {
            v_1 = (double)((double)(v_1) / (double)(2.0));
            k = (double)((double)(k) + (double)(1.0));
        }
        while ((double)(v_1) < (double)(1.0)) {
            v_1 = (double)((double)(v_1) * (double)(2.0));
            k = (double)((double)(k) - (double)(1.0));
        }
        double z_1 = (double)((double)(((double)(v_1) - (double)(1.0))) / (double)(((double)(v_1) + (double)(1.0))));
        double zpow_1 = (double)(z_1);
        double sum_1 = (double)(z_1);
        long i_1 = 3L;
        while ((long)(i_1) <= 9L) {
            zpow_1 = (double)((double)((double)(zpow_1) * (double)(z_1)) * (double)(z_1));
            sum_1 = (double)((double)(sum_1) + (double)((double)(zpow_1) / (double)((((Number)(i_1)).doubleValue()))));
            i_1 = (long)((long)(i_1) + 2L);
        }
        double ln2_1 = (double)(0.6931471805599453);
        return (double)(k) + (double)((double)((double)(2.0) * (double)(sum_1)) / (double)(ln2_1));
    }

    static TextCounts analyze_text(String text) {
        java.util.Map<String,Long> single = ((java.util.Map<String,Long>)(new java.util.LinkedHashMap<String, Long>()));
        java.util.Map<String,Long> double_1 = ((java.util.Map<String,Long>)(new java.util.LinkedHashMap<String, Long>()));
        long n_1 = (long)(_runeLen(text));
        if ((long)(n_1) == 0L) {
            return new TextCounts(single, double_1);
        }
        String last_1 = _substr(text, (int)((long)((long)(n_1) - 1L)), (int)((long)(n_1)));
        if (single.containsKey(last_1)) {
single.put(last_1, (long)((long)(((long)(single).getOrDefault(last_1, 0L))) + 1L));
        } else {
single.put(last_1, 1L);
        }
        String first_1 = _substr(text, (int)(0L), (int)(1L));
        String pair0_1 = " " + first_1;
double_1.put(pair0_1, 1L);
        long i_3 = 0L;
        while ((long)(i_3) < (long)((long)(n_1) - 1L)) {
            String ch_1 = _substr(text, (int)((long)(i_3)), (int)((long)((long)(i_3) + 1L)));
            if (single.containsKey(ch_1)) {
single.put(ch_1, (long)((long)(((long)(single).getOrDefault(ch_1, 0L))) + 1L));
            } else {
single.put(ch_1, 1L);
            }
            String seq_1 = _substr(text, (int)((long)(i_3)), (int)((long)((long)(i_3) + 2L)));
            if (double_1.containsKey(seq_1)) {
double_1.put(seq_1, (long)((long)(((long)(double_1).getOrDefault(seq_1, 0L))) + 1L));
            } else {
double_1.put(seq_1, 1L);
            }
            i_3 = (long)((long)(i_3) + 1L);
        }
        return new TextCounts(single, double_1);
    }

    static long round_to_int(double x) {
        if ((double)(x) < (double)(0.0)) {
            return ((Number)(((double)(x) - (double)(0.5)))).intValue();
        }
        return ((Number)(((double)(x) + (double)(0.5)))).intValue();
    }

    static void calculate_entropy(String text) {
        TextCounts counts = analyze_text(text);
        String alphas_1 = " abcdefghijklmnopqrstuvwxyz";
        long total1_1 = 0L;
        for (String ch : counts.single.keySet()) {
            total1_1 = (long)((long)(total1_1) + (long)(((long)(counts.single).getOrDefault(ch, 0L))));
        }
        double h1_1 = (double)(0.0);
        long i_5 = 0L;
        while ((long)(i_5) < (long)(_runeLen(alphas_1))) {
            String ch_3 = _substr(alphas_1, (int)((long)(i_5)), (int)((long)((long)(i_5) + 1L)));
            if (counts.single.containsKey(ch_3)) {
                double prob_1 = (double)((double)((((double)(counts.single).getOrDefault(ch_3, 0L)))) / (double)((((Number)(total1_1)).doubleValue())));
                h1_1 = (double)((double)(h1_1) + (double)((double)(prob_1) * (double)(log2((double)(prob_1)))));
            }
            i_5 = (long)((long)(i_5) + 1L);
        }
        double first_entropy_1 = (double)(-h1_1);
        System.out.println(_p(round_to_int((double)(first_entropy_1))) + ".0");
        long total2_1 = 0L;
        for (String seq : counts.double_.keySet()) {
            total2_1 = (long)((long)(total2_1) + (long)(((long)(counts.double_).getOrDefault(seq, 0L))));
        }
        double h2_1 = (double)(0.0);
        long a0_1 = 0L;
        while ((long)(a0_1) < (long)(_runeLen(alphas_1))) {
            String ch0_1 = _substr(alphas_1, (int)((long)(a0_1)), (int)((long)((long)(a0_1) + 1L)));
            long a1_1 = 0L;
            while ((long)(a1_1) < (long)(_runeLen(alphas_1))) {
                String ch1_1 = _substr(alphas_1, (int)((long)(a1_1)), (int)((long)((long)(a1_1) + 1L)));
                String seq_3 = ch0_1 + ch1_1;
                if (counts.double_.containsKey(seq_3)) {
                    double prob_3 = (double)((double)((((double)(counts.double_).getOrDefault(seq_3, 0L)))) / (double)((((Number)(total2_1)).doubleValue())));
                    h2_1 = (double)((double)(h2_1) + (double)((double)(prob_3) * (double)(log2((double)(prob_3)))));
                }
                a1_1 = (long)((long)(a1_1) + 1L);
            }
            a0_1 = (long)((long)(a0_1) + 1L);
        }
        double second_entropy_1 = (double)(-h2_1);
        System.out.println(_p(round_to_int((double)(second_entropy_1))) + ".0");
        double diff_1 = (double)((double)(second_entropy_1) - (double)(first_entropy_1));
        System.out.println(_p(round_to_int((double)(diff_1))) + ".0");
    }
    public static void main(String[] args) {
        text1 = "Behind Winston's back the voice " + "from the telescreen was still " + "babbling and the overfulfilment";
        calculate_entropy(text1);
        text3 = "Had repulsive dashwoods suspicion sincerity but advantage now him. " + "Remark easily garret nor nay.  Civil those mrs enjoy shy fat merry. " + "You greatest jointure saw horrible. He private he on be imagine " + "suppose. Fertile beloved evident through no service elderly is. Blind " + "there if every no so at. Own neglected you preferred way sincerity " + "delivered his attempted. To of message cottage windows do besides " + "against uncivil.  Delightful unreserved impossible few estimating " + "men favourable see entreaties. She propriety immediate was improving. " + "He or entrance humoured likewise moderate. Much nor game son say " + "feel. Fat make met can must form into gate. Me we offending prevailed " + "discovery.";
        calculate_entropy(text3);
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _substr(String s, int i, int j) {
        int len = _runeLen(s);
        if (i < 0) i = 0;
        if (j > len) j = len;
        if (i > j) i = j;
        int start = s.offsetByCodePoints(0, i);
        int end = s.offsetByCodePoints(0, j);
        return s.substring(start, end);
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/transpiler/FS/maths/entropy.fs
// Generated 2025-08-16 14:41 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.15g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
type TextCounts = {
    mutable _single: System.Collections.Generic.IDictionary<string, int>
    mutable _double: System.Collections.Generic.IDictionary<string, int>
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
open System.Collections.Generic

let rec log2 (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        let mutable k: float = 0.0
        let mutable v: float = x
        while v >= 2.0 do
            v <- v / 2.0
            k <- k + 1.0
        while v < 1.0 do
            v <- v * 2.0
            k <- k - 1.0
        let z: float = (v - 1.0) / (v + 1.0)
        let mutable zpow: float = z
        let mutable sum: float = z
        let mutable i: int = 3
        while i <= 9 do
            zpow <- (zpow * z) * z
            sum <- sum + (zpow / (float i))
            i <- i + 2
        let ln2: float = 0.6931471805599453
        __ret <- k + ((2.0 * sum) / ln2)
        raise Return
        __ret
    with
        | Return -> __ret
and analyze_text (text: string) =
    let mutable __ret : TextCounts = Unchecked.defaultof<TextCounts>
    let mutable text = text
    try
        let mutable _single: System.Collections.Generic.IDictionary<string, int> = _dictCreate []
        let mutable _double: System.Collections.Generic.IDictionary<string, int> = _dictCreate []
        let n: int = String.length (text)
        if n = 0 then
            __ret <- { _single = _single; _double = _double }
            raise Return
        let last: string = _substring text (n - 1) n
        if _single.ContainsKey(last) then
            _single <- _dictAdd (_single) (string (last)) ((_dictGet _single ((string (last)))) + 1)
        else
            _single <- _dictAdd (_single) (string (last)) (1)
        let first: string = _substring text 0 1
        let pair0: string = " " + first
        _double <- _dictAdd (_double) (string (pair0)) (1)
        let mutable i: int = 0
        while i < (n - 1) do
            let ch: string = _substring text i (i + 1)
            if _single.ContainsKey(ch) then
                _single <- _dictAdd (_single) (string (ch)) ((_dictGet _single ((string (ch)))) + 1)
            else
                _single <- _dictAdd (_single) (string (ch)) (1)
            let seq: string = _substring text i (i + 2)
            if _double.ContainsKey(seq) then
                _double <- _dictAdd (_double) (string (seq)) ((_dictGet _double ((string (seq)))) + 1)
            else
                _double <- _dictAdd (_double) (string (seq)) (1)
            i <- i + 1
        __ret <- { _single = _single; _double = _double }
        raise Return
        __ret
    with
        | Return -> __ret
and round_to_int (x: float) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable x = x
    try
        __ret <- if x < 0.0 then (int (x - 0.5)) else (int (x + 0.5))
        raise Return
        __ret
    with
        | Return -> __ret
and calculate_entropy (text: string) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable text = text
    try
        let counts: TextCounts = analyze_text (text)
        let alphas: string = " abcdefghijklmnopqrstuvwxyz"
        let mutable total1: int = 0
        for ch in (counts._single).Keys do
            total1 <- total1 + (_dictGet (counts._single) ((string (ch))))
        let mutable h1: float = 0.0
        let mutable i: int = 0
        while i < (String.length (alphas)) do
            let ch: string = _substring alphas i (i + 1)
            if (counts._single).ContainsKey(ch) then
                let prob: float = (float (_dictGet (counts._single) ((string (ch))))) / (float total1)
                h1 <- h1 + (prob * (log2 (prob)))
            i <- i + 1
        let first_entropy: float = -h1
        ignore (printfn "%s" ((_str (round_to_int (first_entropy))) + ".0"))
        let mutable total2: int = 0
        for seq in (counts._double).Keys do
            total2 <- total2 + (_dictGet (counts._double) ((string (seq))))
        let mutable h2: float = 0.0
        let mutable a0: int = 0
        while a0 < (String.length (alphas)) do
            let ch0: string = _substring alphas a0 (a0 + 1)
            let mutable a1: int = 0
            while a1 < (String.length (alphas)) do
                let ch1: string = _substring alphas a1 (a1 + 1)
                let seq: string = ch0 + ch1
                if (counts._double).ContainsKey(seq) then
                    let prob: float = (float (_dictGet (counts._double) ((string (seq))))) / (float total2)
                    h2 <- h2 + (prob * (log2 (prob)))
                a1 <- a1 + 1
            a0 <- a0 + 1
        let second_entropy: float = -h2
        ignore (printfn "%s" ((_str (round_to_int (second_entropy))) + ".0"))
        let diff: float = second_entropy - first_entropy
        ignore (printfn "%s" ((_str (round_to_int (diff))) + ".0"))
        __ret
    with
        | Return -> __ret
let text1: string = ("Behind Winston's back the voice " + "from the telescreen was still ") + "babbling and the overfulfilment"
ignore (calculate_entropy (text1))
let text3: string = ((((((((("Had repulsive dashwoods suspicion sincerity but advantage now him. " + "Remark easily garret nor nay.  Civil those mrs enjoy shy fat merry. ") + "You greatest jointure saw horrible. He private he on be imagine ") + "suppose. Fertile beloved evident through no service elderly is. Blind ") + "there if every no so at. Own neglected you preferred way sincerity ") + "delivered his attempted. To of message cottage windows do besides ") + "against uncivil.  Delightful unreserved impossible few estimating ") + "men favourable see entreaties. She propriety immediate was improving. ") + "He or entrance humoured likewise moderate. Much nor game son say ") + "feel. Fat make met can must form into gate. Me we offending prevailed ") + "discovery."
ignore (calculate_entropy (text3))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
